File: test/float32_unpack.js
"use strict";

/*global describe:true, it:true */
var assert = require("chai").assert;

/* 32 bit float (not IEEE; nonnormalized mantissa +
   biased exponent) : neeeeeee eeemmmmm mmmmmmmm mmmmmmmm
   Why not IEEE?  It's just not that important here. */

var VQ_FEXP = 10;
var VQ_FMAN = 21;
var VQ_FEXP_BIAS = 768; /* bias toward values smaller than 1. */

function ldexp(x, exp) {
  return x*Math.pow(2,exp);
}

function float32_unpack(val) {
  if(val===2147483647)return(Infinity);
  var mant=val&0x1fffff;
  var sign=val&0x80000000;
  var exp =(val&0x7fe00000)>>>VQ_FMAN;
  if(sign)mant= -mant;
  return(ldexp(mant,exp-(VQ_FMAN-1)-VQ_FEXP_BIAS));
}


function test(src, actual) {
  it(src + " -> " + actual, function() {
    if (actual === Infinity) {
      assert.equal(float32_unpack(src), actual);
    } else {
      assert.closeTo(float32_unpack(src), actual, 1e-6);
    }
  });
}

describe("float32_pack", function() {
  test(-535612621, -1.200000);
  test(-535822336, -1.000000);
  test(-537919488, -0.500000);
  test(1610612736, 0.000000);
  test(1609564160, 0.500000);
  test(1611661312, 1.000000);
  test(1611871027, 1.200000);
  test(2147483647, Infinity);
  test(-2147483648, -0.000000);
});

File: test/float32_unpack.c
#include <stdio.h>
#include <math.h>

/* 32 bit float (not IEEE; nonnormalized mantissa +
   biased exponent) : neeeeeee eeemmmmm mmmmmmmm mmmmmmmm
   Why not IEEE?  It's just not that important here. */

#define VQ_FEXP 10
#define VQ_FMAN 21
#define VQ_FEXP_BIAS 768 /* bias toward values smaller than 1. */

float _float32_unpack(long val){
  double mant=val&0x1fffff;
  int    sign=val&0x80000000;
  long   exp =(val&0x7fe00000L)>>VQ_FMAN;
  if(sign)mant= -mant;
  return(ldexp(mant,exp-(VQ_FMAN-1)-VQ_FEXP_BIAS));
}


int main(void) {
  long src[] = {
    -535612621,
    -535822336,
    -537919488,
    1610612736,
    1609564160,
    1611661312,
    1611871027,
    +2147483647,
    -2147483648,
  };
  int size = sizeof(src) / sizeof(*src);
  for (int i = 0; i < size; i++) {
    printf("testpack(%ld, %f);\n", src[i], _float32_unpack(src[i]));
  }
  return 0;
}
